### PR TITLE
chore(aci): enforce config schema without subclassing

### PR DESCRIPTION
**Pull request by cathteng**
We will have many detectors of different `types` with different config schemas that need to be enforced. Instead of subclassing and using a Django proxy model we can link the config schema to `GroupType` and enforce it via a pre-save signal.

`Workflows` similarly have a config schema to enforce but all of them have the same schema. So we use a different pre-save signal function for it and a hardcoded schema.

### CODE REVIEW

File: src/sentry/incidents/grouptype.py [hunk 1]
```diff
@@ -26,3 +26,4 @@ class MetricAlertFire(GroupType):
     enable_escalation_detection = False
     detector_handler = MetricAlertDetectorHandler
     detector_validator = MetricAlertsDetectorValidator
+    detector_config_schema = {}  # TODO(colleen): update this
```
Comment (cathteng) [inline src/sentry/incidents/grouptype.py:29]: new

File: src/sentry/issues/grouptype.py [hunk 2]
```diff
@@ -174,11 +174,17 @@ class GroupType:
...
 
     def __init_subclass__(cls: type[GroupType], **kwargs: Any) -> None:
         super().__init_subclass__(**kwargs)
         registry.add(cls)
 
+        from sentry.workflow_engine.registry import detector_config_schema_registry
+
+        detector_config_schema = cls.detector_config_schema or {}
+        detector_config_schema_registry.register(cls.type_id)(json.dumps(detector_config_schema))
+
```
Comment (cathteng) [inline src/sentry/issues/grouptype.py:187]: new

File: src/sentry/utils/registry.py [hunk 3]
```diff
@@ -15,9 +15,10 @@ class NoRegistrationExistsError(ValueError):
 
 
 class Registry(Generic[T]):
-    def __init__(self):
+    def __init__(self, enable_reverse_lookup: bool = True):
```
Comment (cathteng) [inline src/sentry/utils/registry.py:18]: needed to do this because there are multiple detector types that could have the same config schema (e.g. empty dictionary)

File: src/sentry/workflow_engine/models/detector.py [hunk 4]
```diff
@@ -105,3 +104,11 @@ def detector_handler(self) -> DetectorHandler | None:
...
         # TODO: Create proper audit log data for the detector, group and conditions
         return {}
+
+
+@receiver(pre_save, sender=Detector)
+def enforce_config_schema(sender, instance: Detector, **kwargs):
+    from sentry.workflow_engine.registry import detector_config_schema_registry
+
+    config_schema = detector_config_schema_registry.get(instance.type)
+    instance.validate_config(json.loads(config_schema))
```
Comment (cathteng) [inline src/sentry/workflow_engine/models/detector.py:114]: new

File: src/sentry/workflow_engine/models/json_config.py [hunk 5]
```diff
@@ -8,13 +7,9 @@
...
 
-    @abstractproperty
-    def CONFIG_SCHEMA(self) -> dict[str, Any]:
-        pass
-
-    def validate_config(self) -> None:
+    def validate_config(self, schema: dict[str, Any]) -> None:
         try:
-            validate(self.config, self.CONFIG_SCHEMA)
+            validate(self.config, schema)
```
Comment (cathteng) [inline src/sentry/workflow_engine/models/json_config.py:12]: updated, instead we pass in the schema instead of having it defined on the class, and use a pre-save signal to call this function

File: src/sentry/workflow_engine/models/workflow.py [hunk 6]
```diff
@@ -60,3 +63,9 @@ def evaluate_trigger_conditions(self, evt: GroupEvent) -> bool:
 
         evaluation, _ = evaluate_condition_group(self.when_condition_group, evt)
         return evaluation
+
+
+@receiver(pre_save, sender=Workflow)
+def enforce_config_schema(sender, instance: Workflow, **kwargs):
+    config_schema = instance.CONFIG_SCHEMA
+    instance.validate_config(config_schema)
```
Comment (cathteng) [inline src/sentry/workflow_engine/models/workflow.py:71]: new

File: src/sentry/workflow_engine/registry.py [hunk 7]
```diff
@@ -6,3 +6,7 @@
 data_source_type_registry = Registry[type[DataSourceTypeHandler]]()
 condition_handler_registry = Registry[DataConditionHandler[Any]]()
 action_handler_registry = Registry[ActionHandler]()
+
+detector_config_schema_registry = Registry[str](
+    enable_reverse_lookup=False
+)  # json dump, allow duplicate values
```
Comment (cathteng) [inline src/sentry/workflow_engine/registry.py:12]: new

File: tests/sentry/workflow_engine/models/test_json_config_base.py
Comment (cathteng) [inline tests/sentry/workflow_engine/models/test_json_config_base.py:1]: new

**Review by cathteng — COMMENTED** [review]
i needed to update so many tests just to make sure it worked. but the `type_id` in the `Detector` model should be changed to an integer that is one of the registered group type ids, and the serializers and validators need to be updated. doesn't have to be in this PR since i did it hastily

File: src/sentry/issues/grouptype.py [hunk 8]
```diff
@@ -560,12 +569,14 @@ class MonitorIncidentType(GroupType):
...
     # This is deprecated, only kept around for it's type_id
     type_id = 4002
+    slug = "monitor_check_in_timeout"
 
 
 @dataclass(frozen=True)
 class MonitorCheckInMissed(MonitorIncidentType):
     # This is deprecated, only kept around for it's type_id
     type_id = 4003
+    slug = "monitor_check_in_missed"
```
Comment (cathteng) [inline src/sentry/issues/grouptype.py:579]: added slugs in order to not have collisions with the parent slug upon registration.

File: src/sentry/issues/grouptype.py [hunk 9]
```diff
@@ -174,11 +174,20 @@ class GroupType:
...
     def __init_subclass__(cls: type[GroupType], **kwargs: Any) -> None:
         super().__init_subclass__(**kwargs)
         registry.add(cls)
 
+        from sentry.workflow_engine.registry import detector_config_schema_registry
+
+        detector_config_schema = cls.detector_config_schema or {}
+        if (
+            cls.slug not in detector_config_schema_registry.registrations
+        ):  # TODO(cathy): remove after updating getsentry test with patch
```
Comment (cathteng) [inline src/sentry/issues/grouptype.py:188]: getsentry tests are failing because one of the tests is not patching the new registry. will need to update that before removing this

File: src/sentry/utils/registry.py (hunk 3 above)
Comment (cathteng) [inline src/sentry/utils/registry.py:18]: needed to add this to disable reverse lookup for the new registry. there may be multiple detectors with the same config schema (e.g. empty config), and enabling reverse lookup means keys AND values in the registry must be unique.

File: tests/sentry/workflow_engine/processors/test_detector.py [hunk 10]
```diff
@@ -121,9 +121,16 @@ def test_state_results_multi_group(self, mock_produce_occurrence_to_kafka):
         )
 
     def test_no_issue_type(self):
-        detector = self.create_detector(type="invalid slug")
```
Comment (cathteng) [inline tests/sentry/workflow_engine/processors/test_detector.py:124]: with this PR we can no longer create a detector with an invalid slug, so substituting with creating a valid detector and mocking that the GroupType gets deleted

File: src/sentry/workflow_engine/models/detector.py [hunk 11]
```diff
@@ -105,3 +103,11 @@ def detector_handler(self) -> DetectorHandler | None:
     def get_audit_log_data(self) -> dict[str, Any]:
         # TODO: Create proper audit log data for the detector, group and conditions
         return {}
+
+
+@receiver(pre_save, sender=Detector)
+def enforce_config_schema(sender, instance: Detector, **kwargs):
+    from sentry.workflow_engine.registry import detector_config_schema_registry
+
+    config_schema = detector_config_schema_registry.get(instance.type)
```
Comment (cathteng) [inline src/sentry/workflow_engine/models/detector.py:112]: this also enforces that the type is a valid GroupType, because we register all GroupTypes to the registry

File: src/sentry/issues/grouptype.py [hunk 12]
```diff
@@ -174,11 +174,20 @@ class GroupType:
     notification_config: NotificationConfig = NotificationConfig()
     detector_handler: type[DetectorHandler] | None = None
     detector_validator: type[BaseGroupTypeDetectorValidator] | None = None
+    detector_config_schema: dict[str, Any] | None = None
```
Comment (wedamija) [inline src/sentry/issues/grouptype.py:177]: It's starting to feel like we just need something like `DetectorConfig` that encapsulates all of these. Probably something for a separate pr

File: src/sentry/workflow_engine/models/detector.py (hunk 11 above)
Comment (wedamija) [inline src/sentry/workflow_engine/models/detector.py:112]: Do we actually need the registry, or since we have the detector, should we use `instance.group_type.detector_config_schema`?

File: src/sentry/workflow_engine/models/workflow.py [hunk 13]
```diff
@@ -36,7 +38,8 @@ class Workflow(DefaultFieldsModel, OwnerModel, JSONConfigBase):
 
     @property
     def CONFIG_SCHEMA(self) -> dict[str, Any]:
-        raise NotImplementedError('Subclasses must define a "CONFIG_SCHEMA" attribute')
+        # TODO: fill in
+        return {}
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/workflow.py:42]: Is the plan to use subclasses here? Won't we run into the same problem we proxy models?

File: src/sentry/issues/grouptype.py (hunk 8 above)
Comment (wedamija) [reply src/sentry/issues/grouptype.py:579]: I'm not sure if we should do this... I don't know if the slugs are stored anywhere that this would break things

File: src/sentry/utils/registry.py (hunk 3 above)
Comment (wedamija) [inline src/sentry/utils/registry.py:18]: We might not need these changes if we don't actually need the registry, but maybe they're a good addition anyway? Could be split into a separate pr if not required here

File: src/sentry/issues/grouptype.py (hunk 8 above)
Comment (cathteng) [reply src/sentry/issues/grouptype.py:579]: @wedamija is using the `type_id` better?

File: src/sentry/workflow_engine/models/workflow.py (hunk 13 above)
Comment (cathteng) [reply src/sentry/workflow_engine/models/workflow.py:42]: there isn't a type on workflow so all of them _should_ have the same config schema

File: src/sentry/issues/grouptype.py (hunk 8 above)
Comment (cathteng) [reply src/sentry/issues/grouptype.py:579]: edit: no registry means i don't need to add this

File: src/sentry/utils/registry.py (hunk 3 above)
Comment (cathteng) [reply src/sentry/utils/registry.py]: took it out for now

File: src/sentry/workflow_engine/models/workflow.py (hunk 13 above)
Comment (wedamija) [reply src/sentry/workflow_engine/models/workflow.py:42]: I'm not too clear on what these configs are going to be... If they're consistent for all Workflows this is fine though

File: src/sentry/workflow_engine/models/detector.py [hunk 14]
```diff
@@ -105,3 +102,12 @@ def detector_handler(self) -> DetectorHandler | None:
...
         return {}
+
+
+@receiver(pre_save, sender=Detector)
+def enforce_config_schema(sender, instance: Detector, **kwargs):
+    group_type = instance.group_type
+    if not group_type:
+        raise ValueError(f"No group type found with type {instance.type}")
+    config_schema = group_type.detector_config_schema
+    instance.validate_config(config_schema)
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/detector.py:113]: We should probably also check that `detector_config_schema` is set too

File: src/sentry/workflow_engine/models/workflow.py (hunk 6 above)
Comment (wedamija) [inline src/sentry/workflow_engine/models/workflow.py:71]: Nit: Could just one line this

File: src/sentry/workflow_engine/models/workflow.py [hunk 15]
```diff
@@ -36,7 +38,8 @@ class Workflow(DefaultFieldsModel, OwnerModel, JSONConfigBase):
 
     @property
     def CONFIG_SCHEMA(self) -> dict[str, Any]:
```
Comment (wedamija) [inline src/sentry/workflow_engine/models/workflow.py:40]: Nit: Not related to this PR, but we don't usually use all caps for class properties or methods. Usually saved for constants at the module level

**Review by wedamija — APPROVED** [review]

File: src/sentry/workflow_engine/models/detector.py (hunk 14 above)
Comment (cathteng) [reply src/sentry/workflow_engine/models/detector.py:113]: this is set to an empty dict by default in GroupType, do we need anything else?

Comment (wedamija) [reply src/sentry/workflow_engine/models/detector.py:113]: Oh, I guess that's fine... One problem with defaulting it to an empty dict is that if people forget to add it then they won't get an error, but this works as is
